File: Double Doors Regional P6 - Tudor Learns More DDR.cpp
// https://dmoj.ca/problem/ddrp6
// direct translation of Python code, check that version for explanation

#include <bits/stdc++.h>
#define int long long
#define matrix vector<vector<int>>

using namespace std;

const int MOD = 1e9 + 7;

matrix multiply(const matrix& m2, const matrix& m1) {
    int n = m1.size();
    matrix res(n, vector<int>(n));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            int tot = 0;
            for (int k = 0; k < n; k++) {
                tot += m1[k][j] * m2[i][k];  // j -> k -> i
                tot %= MOD;
            }
            res[i][j] = tot;
        }
    }
    return res;
}

matrix matrix_exp(matrix base, int p) {
    int n = base.size();
    matrix res(n, vector<int>(n));  // identity matrix
    for (int i = 0; i < n; i++)
        res[i][i] = 1;
    while (p > 0) {
        if (p % 2 == 1)
            res = multiply(base, res);
        base = multiply(base, base);
        p /= 2;
    }
    return res;
}

int encode(const string &s) {
    int res = 0;
    int pow4[3] = {1, 4, 16};
    for (int i = 0; i < (int)s.size(); i++) {
        res += (s[i] - '0') * pow4[i];
    }
    return res;
}

signed main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    if (n < 10){  // precompute small values to avoid edge cases
        int precomp[] = {-1, 4, 16, 64, 248, 968, 3776, 14728, 57448, 224080, 874040};
        cout << precomp[n] << "\n";
        return 0;
    }
    n -= 3;

    vector<string> states = {
        "000", "001", "002", "003", "010", "011", "012", "013", "020", "021", "022", "023", "030", "031", "032", "033",
        "100", "101", "102", "103", "110", "111", "112", "113", "120", "121", "122", "123", "130", "131", "132", "133",
        "200", "201", "202", "203", "210", "211", "212", "213", "220", "221", "222", "223", "230", "231", "232", "233",
        "300", "301", "302", "303", "310", "311", "312", "313", "320", "321", "322", "323", "330", "331", "332", "333"
    };

    matrix mat(64, vector<int>(64, 0));
    for (const string &cur : states) {
        for (char added : string("0123")) {
            string nxt = cur + added;
            if (string("0123012,0321032").find(nxt) != string::npos) {
                continue;
            }
            nxt = nxt.substr(nxt.size() - 3);
            mat[encode(cur)][encode(nxt)] += 1;
        }
    }

    matrix transition = matrix_exp(mat, n);
    matrix start(64, vector<int>(64, 0));
    for (const string &cur : states) {
        start[encode(cur)][0] += 1;
    }

    matrix res = multiply(transition, start);
    int ans = 0;
    for (const auto &row : res) {
        for (int v : row) {
            ans = (ans + v) % MOD;
        }
    }
    cout << ans << "\n";

    return 0;
}

File: Double Doors Regional P6 - Tudor Learns More DDR.py
# TLE IN PYTHON, CHECK C++ CODE
#
# https://dmoj.ca/problem/ddrp6
# Optimize linear DP transitions by representing them with a matrix and using matrix exponentiation
# The previous 3 moves affect what the next one can be
#
# TC: O(64^3 * log(N))

import sys


def multiply(m2, m1):
    n = len(m1)
    m3 = [[0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):  # j -> idx -> i
            m3[i][j] = sum(m1[idx][j] * m2[i][idx] for idx in range(n)) % MOD
    return m3


def matrix_exp(original_mat, p):
    n = len(original_mat)
    base = [row.copy() for row in original_mat]
    res = [[0] * n for _ in range(n)]  # identity matrix
    for i in range(n):
        res[i][i] = 1
    while p > 0:
        if p % 2 == 1:
            res = multiply(base, res)
        base = multiply(base, base)
        p //= 2
    return res


def encode(s):
    res = 0
    for i in range(len(s)):
        res += int(s[i]) * (4 ** i)
    return res


MOD = 10 ** 9 + 7
n = int(input())

if n < 10:  # precompute small values to avoid edge cases
    precomp = [None, 4, 16, 64, 248, 968, 3776, 14728, 57448, 224080, 874040]
    print(precomp[n])
    sys.exit()

n -= 3

states = ['000', '001', '002', '003', '010', '011', '012', '013', '020', '021', '022', '023', '030', '031', '032',
          '033', '100', '101', '102', '103', '110', '111', '112', '113', '120', '121', '122', '123', '130', '131',
          '132', '133', '200', '201', '202', '203', '210', '211', '212', '213', '220', '221', '222', '223', '230',
          '231', '232', '233', '300', '301', '302', '303', '310', '311', '312', '313', '320', '321', '322', '323',
          '330', '331', '332', '333']

mat = [[0] * 64 for _ in range(64)]
for cur in states:  # encode valid transitions in matrix
    for added in "0123":
        nxt = cur + added
        if nxt in "0123012,0321032":  # disallowed
            continue
        nxt = nxt[-3:]
        mat[encode(cur)][encode(nxt)] += 1

transition = matrix_exp(mat, n)  # transition `n` times

start = [[0] * 64 for _ in range(64)]  # apply transitions to starting state
for cur in states:
    start[encode(cur)][0] += 1
res = multiply(transition, start)

print(sum(sum(row) for row in res) % MOD)

# # Brute force DP code, O(N * large constant)
# bad = "0123012,0321032"
#
# n = int(input())
#
# ending = Counter()  # number of sequences ending with ending[i]
# ending[""] = 1
# for i in range(n):
#     state = Counter()
#     for cur in "0123":
#         for prev in ending:
#             nxt = prev + cur
#             nxt = nxt[-4:]  # only check last 4 char
#             if len(nxt) != 4 or nxt not in bad:  # valid transition
#                 nxt = nxt[-3:]  # only store last 3 chars
#                 state[nxt] += ending[prev]
#             state[nxt] %= MOD
#
#     ending = state
#
# # print(ending)
# print(sum(ending.values()) % MOD)
